compiler/executable/match_/planner/vertex/constraint.rs

```diff
@@ -1311,6 +1321,36 @@ impl Costed for PlaysPlanner<'_> {
         _: Option<Direction>,
         _: &Graph<'_>,
     ) -> Result<(Cost, CostMetaData), QueryPlanningError> {
-        Ok((Cost::in_mem_complex_with_ratio(1.0), CostMetaData::Direction(Direction::Canonical)))
+        Ok((Cost::in_mem_complex_with_ratio(1.0), CostMetaData::Direction(Direction::Canonical, None)))
+    }
+}
+
+fn pick_join_var(
+    direction: Direction,
+    is_from_bound: bool,
+    is_to_bound: bool,
+    from_var: VariableVertexId,
+    to_var: VariableVertexId,
+) -> Option<VariableVertexId> {
+    if is_from_bound && is_to_bound {
+        None
+    } else {
+        let join_var = match direction {
+            Direction::Canonical => {
+                if is_from_bound {
+                    to_var
+                } else {
+                    from_var
+                }
+            }
+            Direction::Reverse => {
+                if is_to_bound {
+                    from_var
+                } else {
+                    to_var
+                }
+            }
+        };
+        Some(join_var)
     }
```

Comment:
Optionally reduce rightward drift:
```suggestion
    if is_from_bound && is_to_bound {
        return None;
    }
    match direction {
        Direction::Canonical if is_from_bound => Some(to_var),
        Direction::Canonical => Some(from_var),
        Direction::Reverse if is_to_bound => Some(from_var),
        Direction::Reverse => Some(to_var),
    }
```